modules/loader/src/main/scala/com/snowplowanalytics/snowplow/postgres/loader/Main.scala

```diff
@@ -69,27 +71,63 @@ object Main extends IOApp {
   ): IO[ExitCode] = {
       implicit val db: DB[IO] = DB.interpreter[IO](cli.iglu.resolver, xa, cli.config.output.good.schema)
       for {
-        now <- TimeUtils.now[IO]
         _ <- cli.config.purpose match {
           case Purpose.Enriched       => utils.prepare[IO](cli.config.output.good.schema, xa)
           case Purpose.SelfDescribing => IO.unit
         }
-        _ <- env.source
-          .through(parsePayload(env.getPayload, cli.config.purpose, now))
-          .through(sinkAll(env.sinkPipe(xa), Sink.sinkResult(state, cli.iglu, processor, env.badRowSink)))
-          .through(env.checkpointer)
-          .compile
-          .drain
+        _ <- runWithShutdown {
+            env.source
+              .through(parsePayload(env.getPayload, cli.config.purpose))
+              .through(sinkAll(env.sinkPipe(xa), Sink.sinkResult(state, cli.iglu, processor, env.badRowSink)))
+          }(_.through(env.checkpointer))
       } yield ExitCode.Success
   }
 
-  private def parsePayload[A](getPayload: A => Either[BadRow, String], purpose: Purpose, now: Instant): Pipe[IO, A, (A, Either[BadRow, Data])] =
-    _.map { record =>
+  /**
+   * This is the machinery needed to make sure outstanding records are checkpointed before the app
+   * terminates
+   *
+   * The stream runs on a separate fiber.  We call `join` on the fiber, so any runtime error on the
+   * stream causes the app to shut down.
+   *
+   * We use a queue as a level of indirection between the soure and the sink. When we receive a
+   * SIGINT then we terminate the fiber by pushing a `None` to the queue.
+   *
+   * We call `interruptWhen` on the `queue.dequeue` because it causes all outstanding records to
```

Comment:
TODO: Update this scaladoc.  A previous implementation used `interruptWhen` but the latest implementation uses a NoneTerminatedQueue instead.  But in both cases we terminate the queue rather than the source., to avoid shutting down the kinesis scheduler.